docs/design/proposals/rbd-volume-healer.md

```diff
@@ -0,0 +1,102 @@
+# RBD NBD VOLUME HEALER
+
+- [RBD NBD VOLUME HEALER](#rbd-nbd-volume-healer)
+  - [Rbd Nbd](#rbd-nbd)
+    - [Advantages of userspace mounters](#advantages-of-userspace-mounters)
+    - [Side effects of userspace mounters](#side-effects-of-userspace-mounters)
+  - [Volume Healer](#volume-healer)
+    - [More thoughts](#more-thoughts)
+
+## Rbd nbd
+
+The rbd CSI plugin will provision new rbd images and attach and mount those
+to workloads. Currently, the default mounter is krbd, which uses the kernel
+rbd driver to mount the rbd images onto the application pod. Here on
+at Ceph-CSI we will also have a userspace way of mounting the rbd images,
+via rbd-nbd.
+
+[Rbd-nbd](https://docs.ceph.com/en/latest/man/8/rbd-nbd/) is a client for
+RADOS block device (rbd) images like the existing rbd kernel module. It
+will map an rbd image to an nbd (Network Block Device) device, allowing
+access to it as a regular local block device.
+
+![csi-rbd-nbd](./images/csi-rbd-nbd.svg)
+
+It’s worth making a note that the rbd-nbd processes will run on the
+client-side, which is inside the `csi-rbdplugin` node plugin.
+
+### Advantages of userspace mounters
```

Comment:
This chapter trades stability (of kernel) for features (of rbd-nbd).

What a filesystem (xfs, ext4) does when rbd-nbd is not running at the time an application issues `write()` there? How long do they wait for any user space daemon to come back? What is the probability of loosing data and/or corrupting the filesystem?

IMO all write errors lead to remounting the volume as read-only and requiring remount (to replay logs and potentially fix the filesystem). Is this going to happen basically randomly when rbd-nbd pods are deleted and take too long to restart?

A block device managed by user space will be much more error prone, is it really worth the trade?

Reply:
Yes our CI already caught some of these problems when we initially injected few random tests. We are in discussion with the ceph core team if it is a good idea to freeze the filesystem as part of detach operation with the help of rbd-nbd quience hooks. Thereafter attempting to perform any modifications to the frozen fs will be blocked waiting until the fs is unfrozen which will be achieved by attaching the process back.

Reply:
@pkalever please add a link to the discussion (ideally tracked in a ceph-csi issue)

Reply:
> What a filesystem (xfs, ext4) does when rbd-nbd is not running at the time an application issues write() there? How long do they wait for any user space daemon to come back?

When rbd-nbd process is terminated, the IO will wait until the nbd NBD_ATTR_TIMEOUT is reached. NBD_ATTR_TIMEOUT can be set by `--io-timeout` option of rbd-nbd tool, It specifies how long the IO should wait to get handled before bailing out. When we set the value to zero, it should keep the requeueing logic enabled, Specifying `--io-timeout` 0 tells the nbd driver to not abort the request and instead see if it can be restarted on another socket.

We can set NBD_ATTR_DEAD_CONN_TIMEOUT which controls, how long the device should be held waiting for the userspace process to come back to life. This can be set with rbd-nbd option `--reattach-timeout`

> What is the probability of loosing data and/or corrupting the filesystem?
> IMO all write errors lead to remounting the volume as read-only and requiring remount (to replay logs and potentially fix the filesystem). Is this going to happen basically randomly when rbd-nbd pods are deleted and take too long to restart?

Data/Metadata loss will probably happen only when the nodeplugin pod doesn't come back in time after a restart/upgrade, again this should be extremely rare IMHO.

> A block device managed by userspace will be much more error prone, is it really worth the trade?

I agree, but we can make it work better for sure, for example, gluster-block storage (openshift container storage 3) is based on tcmu-runner userspace block service and our users are happy with it.